{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from time import sleep\n",
    "from bs4 import BeautifulSoup\n",
    "from datetime import date\n",
    "from datetime import datetime\n",
    "from dateutil import parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "regex = '\"slug\":\"(.*?)\".*?\"link\":\"(.*?)\".*?\"title\":{\"rendered\":\"(.*?)\"}.*?\"content\":{\"rendered\":\"(.*?)\"}'\n",
    "\n",
    "ppp = 100\n",
    "pages = round(754/100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/matthewwithanm/python-markdownify\n",
    "from bs4 import BeautifulSoup, NavigableString\n",
    "import re\n",
    "import six\n",
    "\n",
    "convert_heading_re = re.compile(r'convert_h(\\d+)')\n",
    "line_beginning_re = re.compile(r'^', re.MULTILINE)\n",
    "whitespace_re = re.compile(r'[\\r\\n\\s\\t ]+')\n",
    "\n",
    "# Heading styles\n",
    "ATX = 'atx'\n",
    "ATX_CLOSED = 'atx_closed'\n",
    "UNDERLINED = 'underlined'\n",
    "SETEXT = UNDERLINED\n",
    "\n",
    "def escape(text):\n",
    "    if not text:\n",
    "        return ''\n",
    "    return text.replace('_', r'\\_')\n",
    "\n",
    "\n",
    "def chomp(text):\n",
    "    \"\"\"\n",
    "    If the text in an inline tag like b, a, or em contains a leading or trailing\n",
    "    space, strip the string and return a space as suffix of prefix, if needed.\n",
    "    This function is used to prevent conversions like\n",
    "        <b> foo</b> => ** foo**\n",
    "    \"\"\"\n",
    "    prefix = ' ' if text and text[0] == ' ' else ''\n",
    "    suffix = ' ' if text and text[-1] == ' ' else ''\n",
    "    text = text.strip()\n",
    "    return (prefix, suffix, text)\n",
    "\n",
    "\n",
    "def _todict(obj):\n",
    "    return dict((k, getattr(obj, k)) for k in dir(obj) if not k.startswith('_'))\n",
    "\n",
    "\n",
    "class MarkdownConverter(object):\n",
    "    class DefaultOptions:\n",
    "        strip = None\n",
    "        convert = None\n",
    "        autolinks = True\n",
    "        heading_style = UNDERLINED\n",
    "        bullets = '*+-'  # An iterable of bullet types.\n",
    "\n",
    "    class Options(DefaultOptions):\n",
    "        pass\n",
    "\n",
    "    def __init__(self, **options):\n",
    "        # Create an options dictionary. Use DefaultOptions as a base so that\n",
    "        # it doesn't have to be extended.\n",
    "        self.options = _todict(self.DefaultOptions)\n",
    "        self.options.update(_todict(self.Options))\n",
    "        self.options.update(options)\n",
    "        if self.options['strip'] is not None and self.options['convert'] is not None:\n",
    "            raise ValueError('You may specify either tags to strip or tags to'\n",
    "                             ' convert, but not both.')\n",
    "\n",
    "    def convert(self, html):\n",
    "        soup = BeautifulSoup(html, 'html.parser')\n",
    "        return self.process_tag(soup, children_only=True)\n",
    "\n",
    "    def process_tag(self, node, children_only=False):\n",
    "        text = ''\n",
    "\n",
    "        # Convert the children first\n",
    "        for el in node.children:\n",
    "            if isinstance(el, NavigableString):\n",
    "                text += self.process_text(six.text_type(el))\n",
    "            else:\n",
    "                text += self.process_tag(el)\n",
    "\n",
    "        if not children_only:\n",
    "            convert_fn = getattr(self, 'convert_%s' % node.name, None)\n",
    "            if convert_fn and self.should_convert_tag(node.name):\n",
    "                text = convert_fn(node, text)\n",
    "\n",
    "        return text\n",
    "\n",
    "    def process_text(self, text):\n",
    "        return escape(whitespace_re.sub(' ', text or ''))\n",
    "\n",
    "    def __getattr__(self, attr):\n",
    "        # Handle headings\n",
    "        m = convert_heading_re.match(attr)\n",
    "        if m:\n",
    "            n = int(m.group(1))\n",
    "\n",
    "            def convert_tag(el, text):\n",
    "                return self.convert_hn(n, el, text)\n",
    "\n",
    "            convert_tag.__name__ = 'convert_h%s' % n\n",
    "            setattr(self, convert_tag.__name__, convert_tag)\n",
    "            return convert_tag\n",
    "\n",
    "        raise AttributeError(attr)\n",
    "\n",
    "    def should_convert_tag(self, tag):\n",
    "        tag = tag.lower()\n",
    "        strip = self.options['strip']\n",
    "        convert = self.options['convert']\n",
    "        if strip is not None:\n",
    "            return tag not in strip\n",
    "        elif convert is not None:\n",
    "            return tag in convert\n",
    "        else:\n",
    "            return True\n",
    "\n",
    "    def indent(self, text, level):\n",
    "        return line_beginning_re.sub('\\t' * level, text) if text else ''\n",
    "\n",
    "    def underline(self, text, pad_char):\n",
    "        text = (text or '').rstrip()\n",
    "        return '%s\\n%s\\n\\n' % (text, pad_char * len(text)) if text else ''\n",
    "\n",
    "    def convert_a(self, el, text):\n",
    "        prefix, suffix, text = chomp(text)\n",
    "        if not text:\n",
    "            return ''\n",
    "        href = el.get('href')\n",
    "        title = el.get('title')\n",
    "        if self.options['autolinks'] and text == href and not title:\n",
    "            # Shortcut syntax\n",
    "            return '<%s>' % href\n",
    "        title_part = ' \"%s\"' % title.replace('\"', r'\\\"') if title else ''\n",
    "        return '%s[%s](%s%s)%s' % (prefix, text, href, title_part, suffix) if href else text\n",
    "\n",
    "    def convert_b(self, el, text):\n",
    "        return self.convert_strong(el, text)\n",
    "\n",
    "    def convert_blockquote(self, el, text):\n",
    "        return '\\n' + line_beginning_re.sub('> ', text) if text else ''\n",
    "\n",
    "    def convert_br(self, el, text):\n",
    "        return '  \\n'\n",
    "\n",
    "    def convert_em(self, el, text):\n",
    "        prefix, suffix, text = chomp(text)\n",
    "        if not text:\n",
    "            return ''\n",
    "        return '%s*%s*%s' % (prefix, text, suffix)\n",
    "\n",
    "    def convert_hn(self, n, el, text):\n",
    "        style = self.options['heading_style']\n",
    "        text = text.rstrip()\n",
    "        if style == UNDERLINED and n <= 2:\n",
    "            line = '=' if n == 1 else '-'\n",
    "            return self.underline(text, line)\n",
    "        hashes = '#' * n\n",
    "        if style == ATX_CLOSED:\n",
    "            return '%s %s %s\\n\\n' % (hashes, text, hashes)\n",
    "        return '%s %s\\n\\n' % (hashes, text)\n",
    "\n",
    "    def convert_i(self, el, text):\n",
    "        return self.convert_em(el, text)\n",
    "\n",
    "    def convert_list(self, el, text):\n",
    "        nested = False\n",
    "        while el:\n",
    "            if el.name == 'li':\n",
    "                nested = True\n",
    "                break\n",
    "            el = el.parent\n",
    "        if nested:\n",
    "            # remove trailing newline if nested\n",
    "            return '\\n' + self.indent(text, 1).rstrip()\n",
    "        return '\\n' + text + '\\n'\n",
    "\n",
    "    convert_ul = convert_list\n",
    "    convert_ol = convert_list\n",
    "\n",
    "    def convert_li(self, el, text):\n",
    "        parent = el.parent\n",
    "        if parent is not None and parent.name == 'ol':\n",
    "            if parent.get(\"start\"):\n",
    "                start = int(parent.get(\"start\"))\n",
    "            else:\n",
    "                start = 1\n",
    "            bullet = '%s.' % (start + parent.index(el))\n",
    "        else:\n",
    "            depth = -1\n",
    "            while el:\n",
    "                if el.name == 'ul':\n",
    "                    depth += 1\n",
    "                el = el.parent\n",
    "            bullets = self.options['bullets']\n",
    "            bullet = bullets[depth % len(bullets)]\n",
    "        return '%s %s\\n' % (bullet, text or '')\n",
    "\n",
    "    def convert_p(self, el, text):\n",
    "        return '%s\\n\\n' % text if text else ''\n",
    "\n",
    "    def convert_strong(self, el, text):\n",
    "        prefix, suffix, text = chomp(text)\n",
    "        if not text:\n",
    "            return ''\n",
    "        return '%s**%s**%s' % (prefix, text, suffix)\n",
    "\n",
    "    def convert_img(self, el, text):\n",
    "        alt = el.attrs.get('alt', None) or ''\n",
    "        src = el.attrs.get('src', None) or ''\n",
    "        title = el.attrs.get('title', None) or ''\n",
    "        title_part = ' \"%s\"' % title.replace('\"', r'\\\"') if title else ''\n",
    "        return '![%s](%s%s)' % (alt, src, title_part)\n",
    "\n",
    "\n",
    "def markdownify(html, **options):\n",
    "    return MarkdownConverter(**options).convert(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "author_xml = \"<wp:author_login><!\\[CDATA\\[(.*?)\\]\\]>.*<wp:author_display_name><!\\[CDATA\\[(.*?)]]><\\/wp:author_display_name>.*<\\/wp:author>\"\n",
    "\n",
    "authors = {}\n",
    "\n",
    "with open('/Users/kamranahmed/Developer/python/bsr-json-regex/posts/authors.xml', 'r') as xml:\n",
    "    contents = xml.read()\n",
    "    for auth_id, auth_display_name in re.findall(author_xml, contents):\n",
    "        authors[auth_id] = auth_display_name\n",
    "\n",
    "        \n",
    "media_xml = \"<item>[\\s\\S]*?<guid .*?>(.*?)<\\/guid>[\\s\\S]*?<wp:post_parent>(.*?)<\\/wp:post_parent>[\\s\\S]*?<\\/item>\"\n",
    "        \n",
    "media = {}\n",
    "        \n",
    "with open('/Users/kamranahmed/Developer/python/bsr-json-regex/posts/media.xml', 'r') as xml:\n",
    "    contents = xml.read()\n",
    "    for url, parent_id in re.findall(media_xml, contents):\n",
    "        media[parent_id] = url        \n",
    "        \n",
    "regex_xml = \"<item>[\\s\\S]*?<title>(.*?)<\\/title>[\\s\\S]*?<link>(.*?)<\\/link>[\\s\\S]*?<pubDate>(.*?)<\\/pubDate>[\\s\\S]*?<dc:creator><!\\[CDATA\\[(.*?)\\]\\]><\\/dc:creator>[\\s\\S]*?p=(.*?)<\\/guid>[\\s\\S]*?<content:encoded><!\\[CDATA\\[([\\s\\S]*?)\\]\\][\\s\\S]*?<\\/item>\"\n",
    "\n",
    "posts = []\n",
    "\n",
    "with open('/Users/kamranahmed/Developer/python/bsr-json-regex/posts/export.xml', 'r') as xml:\n",
    "    contents = xml.read()\n",
    "    for post in re.findall(regex_xml, contents):\n",
    "        posts.append({})\n",
    "        ID = post[4]\n",
    "        posts[-1][\"id\"] = ID\n",
    "        posts[-1][\"title\"] = markdownify(post[0])\n",
    "        posts[-1][\"link\"] = post[1]\n",
    "        posts[-1][\"date\"] = parser.parse(post[2]).strftime('%d %B %Y %I:%M%p')\n",
    "        posts[-1][\"author\"] = authors[post[3]]\n",
    "        if ID in media:\n",
    "            posts[-1]['featured_image'] = media[ID]\n",
    "        else:\n",
    "            posts[-1]['featured_image'] = ''\n",
    "            \n",
    "        newlines = re.sub('\\n', '<br>', post[5])\n",
    "        hlines = re.sub('<hr />', '-------', newlines)\n",
    "        out = re.sub('<!-- .*? -->', '', hlines)\n",
    "        final = re.sub('\\[caption .*?\\]', '', out)\n",
    "        final = re.sub('\\[\\/caption\\]', '', final)\n",
    "\n",
    "        md = markdownify(final, heading_style=ATX).lstrip()\n",
    "        final = re.sub('\\n\\s*\\n', '<br><br>', md)\n",
    "        \n",
    "        posts[-1][\"markdown\"] = markdownify(final, heading_style=ATX).lstrip()\n",
    "        \n",
    "#         caption = re.sub('\\[caption .*?\\[\\/caption]', '', post[5]\n",
    "#         comments = re.sub('<!-- .*? -->', '', breaks)\n",
    "#         hlines = re.sub('<hr />', '-------', comments)\n",
    "#         md = markdownify(md, heading_style=ATX).lstrip()\n",
    "#         md = newlines = re.sub(r'\\n\\s*\\n ', '\\n\\n', hlines)\n",
    "        \n",
    "# #         newlines = re.sub('\\n\\s*\\n', '', md)\n",
    "        \n",
    "# #         final = re.sub('\\[caption .*?\\[\\/caption]', '', out)\n",
    "# #         final = re.sub('\\n\\s*\\n', '\\n\\n', final)\n",
    "\n",
    "# #         posts[-1][\"markdown\"] = markdownify(hlines, heading_style=ATX).lstrip()\n",
    "#         posts[-1][\"markdown\"] = md\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "***Part I in our COVID-19 series, Swabs and Spit. We take an inside look into how Berkeley scientists have set the bar for COVID-19 testing as the campus community heads back to work.***   \n",
      "  \n",
      "As biologists, most of our days are spent toiling in a tucked-away lab prying at core questions important to biology, such as: How do cells regulate their size? How do organisms respond to pathogens? Why do we age? To answer these questions, we have mastered molecular techniques such as PCR, a way to amplify small amounts of DNA in our samples and feed our curiosities by becoming experts about specific cellular pathways and model organisms.   \n",
      "  \n",
      "Rarely are we called upon to apply this expertise to solve a tangible, widespread problem with immediate implications for all of humanity. ## The Need for Testing  \n",
      "  \n",
      "In early spring, scientists at UC Berkeley started hearing that call. The COVID-19 pandemic highlighted a unique challenge that was emerging all around the world. We needed to increase testing efforts, and fast, but the authorizations to do so are typically long and arduous and certainly not practical for a situation which required a rapid response. Despite these challenges, a team of Berkeley scientists, led by Jennifer Doudna, Fyodor Urnov and Dirk Hockemeyer, implemented a so-called ”pop-up” testing lab in just three weeks. In tandem, the [Innovative Genomics Institute (IGI)](https://innovativegenomics.org/about-us/) also created a platform to provide the campus community with asymptomatic testing—an absolutely necessary service as we return to work and school. I am humbled to be a part of this volunteer team, made of friends and colleagues from various departments, who have come together in the face of uncertainty and what sometimes feels, frankly, like a chaotic world. I am happy to write a series of posts that will cover the immense efforts involved in this undertaking and, hopefully, emphasize the true team endeavor this has been in bringing testing to campus and the surrounding Bay Area community.   \n",
      "  \n",
      "The need for widespread testing became obvious in early- to mid-March, as the World Health Organization (W.H.O.) declared the coronavirus outbreak a pandemic. The infectious disease, COVID-19, caused by a novel coronavirus, SARS-CoV-2, was spreading rapidly around the globe. Just days after the pandemic declaration, Berkeley scientists at the IGI decided to create a pop-up testing lab on campus to meet this need. This effort has been led by IGI Founder and Board President Doudna, who is a world-renowned scientist most notably famous for her pioneering work in CRISPR-based genome editing technologies. The effort is co-lead by Urnov, IGI’s Scientific Director of Technology and Translation, who also has an impressive career pushing the boundaries of genome-engineering technologies and Hockemeyer, associate professor of Cell and Developmental Biology, whose research has paved the way for genome-editing in stem cells. The decision to create a diagnostic testing lab has not only required these three talented scientists, but also the incredible efforts of over 50 students, post-doctoral researchers, and professors, all of whom are volunteering their time. ## The Hurdles of Setting up a Clinical Lab  \n",
      "  \n",
      "The decision to start testing was a no-brainer to the IGI team; though there were many hurdles to become certified for clinical testing. This process typically involves applying for a state license as well as a Clinical Laboratory Improvement Amendments (CLIA) certification, which alone can take up to six weeks to be approved. On top of this, all personnel in the lab need to have a clinical lab scientist (CLS) license, which involves a one-year full-time internship and passing a comprehensive written exam. So how was this possible? How could UC Berkeley scientists set up a clinical testing lab in just three weeks?   \n",
      "  \n",
      "For starters, some of these regulatory requirements were relaxed in the face of the emergency pandemic situation. This allowed the team of scientists to extend a CLIA certification that already existed on campus—at the University Health Services, Tang Center—to temporarily include a pop-up testing facility at the IGI. Additionally, although the IGI testing facility is managed by a formally trained and certified clinical lab scientist, Lisa Martell, an [executive order signed by Governor Gavin Newson](https://www.gov.ca.gov/wp-content/uploads/2020/03/3.12.20-EO-N-25-20-COVID-19.pdf) allowed our testing pipeline to be carried out by academic scientists. These scientists must be trained within the rigorous landscape of diagnostic laboratory regulations which include virtual training sessions to comply with EH&S and HIPAA. All of this has allowed volunteers including myself and many other graduate students and postdocs to work in the testing facility handling patient samples.   \n",
      "  \n",
      "One major challenge at the forefront of creating a SARS-CoV-2 testing facility centered around the need for a reliable testing protocol that could be authorized by the FDA for Emergency Use Authorization (EUA). In mid-March, this challenge became even more difficult with extensive personal protective equipment and [reagent shortages reported across the country](https://www.nytimes.com/2020/03/11/health/coronavirus-testing-shortages.html). These problems were solved, in part, by adapting a COVID-19 test that already had an approved EUA protocol from the FDA. This test involves using pipetting robots, operated by volunteer scientists, to extract genetic material from a sample and amplify genes specific to SARS-CoV-2 virus that may be present in the sample. Detection of viral genes below an amplification threshold would indicate viral presence within that sample. This PCR-based test, marketed by Thermo Fisher Scientific, was a perfect fit not only because of its compatibility with the semi-automated robotics system being set up in the lab, but also because of reagent availability and promising results gathered from pilot experiments. Since several adaptations were made to the protocol in order to make this test more high-throughput and cost effective, the IGI testing facility team were required to apply for a new EUA. For more information on the specifics of the IGI’s process, check out the [published blueprint for a pop-up SARS-CoV-2 testing lab](https://www.nature.com/articles/s41587-020-0583-3).   \n",
      "  \n",
      "Another essential piece of the puzzle was implementing a Laboratory Information Management System (LIMs), to securely track patient samples from the clinic, through the lab, and report results. This was established through a collaboration with three cloud-based software companies: ThirdWave Analytics (TWA), Salesforce, and Mulesoft. TWA responded to an urgent call to build a fully custom system and HIPAA-compliant physician requisitioning portal, with Salesforce licensing support. Mulesoft then offered free services to integrate the system with the California Public Health Department’s reporting infrastructure to enable the fast and automated reporting necessary to track and respond to the pandemic at a state level. This system, built in a matter of weeks, would normally take months to build and was essential in allowing secure clinical testing to start at the IGI on April 6th, just over three weeks after COVID-19 was declared a pandemic.   \n",
      "  \n",
      "![](https://berkeleysciencereview.com/wp-content/uploads/2020/09/IGI\\_P1\\_FINAL\\_Phil\\_edits.png) **Timeline for SARS-CoV-2 pop-up testing lab.** Within one week of the pandemic declaration, Berkeley scientists decided to create a pop-up testing lab on campus. With help from Salesforce, headquartered in San Francisco, the laboratory was set up and took its first samples three weeks after the initial idea. ## The Testing Pipeline  \n",
      "  \n",
      "When I first joined as a volunteer, I was fascinated to learn about how streamlined the whole testing process was. Patient samples arrive from the clinic, go through the testing pipeline, and the results are reported back to the patient in as short as a day or two. While this whole process is securely tracked through the LIMs set up in the testing facility, there are many other parts that go into testing clinical samples. The process generally starts at a clinical partner site where a sample is taken from a patient with a medical equivalent of a long q-tip used to swab the back of the nose and throat. This sample swab immediately goes in a tube with media to inactivate any virus and preserve the integrity of the viral RNA that may be present. The tube containing the sample is then placed in a biohazard bag and brought to the IGI testing facility. There, it encounters the first batch of volunteers at the testing facility—the intake team. Intake involves verifying patient health information, effectively checking to make sure the tube submitted matches the patient it came from, sanitizing the tube in a biosafety hood, and data entry of the sample into the LIMs, allowing the team to track the sample through the testing facility from intake to testing result. This particular process took some time to troubleshoot and perfect; before integration into the LIMs system, intake took much longer. “On week two, it took nearly 4.5 hours to process 110 samples”, recounts Liz O’Brien, a lead postdoc volunteer on the intake team. “Now we can do close to 500 samples in three hours!”   \n",
      "  \n",
      "The samples are then uncapped, with the help of more volunteers, and loaded into a robot that arrays them into a 96-well plate, all while keeping track of which well contained each individual sample. Next, the genetic material from each sample is extracted and used in a PCR reaction to amplify and detect viral genes. If two out of three viral genes are detected, the system recognizes and reports this test as a clinical positive, notifying the clinician. For the rest, where viral genes are not detected in the sample, it is a clinically negative test and the clinician is notified. Currently, the lab’s capacity is around 1,000 tests per day, with a turnaround time of 48 hours, which is much faster than other testing centers statewide where patients might wait four to five days before receiving their results.   \n",
      "  \n",
      "Building this pipeline has been filled with challenges that the team must overcome. Although the lab has been up and running for over four months, these challenges still arise and need to be solved. For example, before the protocol was approved for the use of the pipetting robots, the process of extracting genetic material and running the PCR reaction was done manually. Since approval of the protocol, all of the testing has been semi-automated with robots. While this innovation allows a huge increase in sample capacity, it leaves the lab heavily dependent on the robots working properly. From time to time, they fail, and the lab is on the clock to identify and fix the problem. When this happens, samples are still coming in. The longer it takes to fix the problem, the larger the sample backlog becomes. ## It Takes a Village  \n",
      "  \n",
      "If that sounds like a lot of work, with an endless number of hurdles, that’s because it was! The IGI team took this in stride though and, in just three weeks, the idea of this pop-up lab became a reality when the testing lab received their first samples from the clinic. To date, their efforts in creating a clinical testing laboratory have helped clinicians across the Bay Area administer and report the results of around 30,000 tests, many of which have been for particularly at-risk communities such as the homeless and elderly populations. Equally impressive is the number of volunteers that have stepped up and come together to make this all possible. The efforts of over 50 people, both volunteer and full-time employees, have carried this from an idea to a reality. People like Shana McDevitt, director of the UC Berkeley genomics sequencing core, have put their “normal” roles on hold, working around the clock, quite literally, to get the testing facility off the ground and keep it running smoothly. She likens this effort to that of, “an army working seven days a week, 10 to 12 hours a day”. What’s that old adage? Oh, yes! It takes a village to start a pop-up clinical testing lab.   \n",
      "  \n",
      "*This blog post is just the first of a series of three outlining the efforts of a campus-wide initiative to increase testing capacity to keep our community safe. Keep an eye out for Part II, which will cover how volunteers built a “saliva kiosk” for the rapid testing of students, faculty and staff on campus.*   \n",
      "  \n",
      "-------   \n",
      "  \n",
      "*Phillip Frankino is a graduate student in molecular and cell biology*   \n",
      "  \n",
      "*Designs by Santiago Yori*\n"
     ]
    }
   ],
   "source": [
    "# with open('/Users/kamranahmed/Developer/python/bsr-json-regex/md.txt', 'w+') as f:\n",
    "#     f.write(posts['17781']['markdown'])\n",
    "print(posts[-1]['markdown'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "y = json.dumps(posts, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/kamranahmed/Developer/python/bsr-json-regex/testing.js', 'w', encoding='utf8') as f:\n",
    "    f.write(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/Users/kamranahmed/Developer/js/blog_strapi_gatsby/cms/config/functions/export.js', 'w', encoding='utf8') as json_file:\n",
    "    json.dump(posts, json_file, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = re.sub(\"\\[!\\[\\((.*?)\\)\\]\\(.*?\\)\", r\"![](\\1)\", \"[![(http://3.bp.blogspot.com/-0pp9oQtCdX0/UV8BUtg\\\\_6\\\\_I/AAAAAAAAAMk/OVWkD\\\\_jegeM/s1600/goal.jpg)](http://wp.me/p1xHew-25U)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'![](http://3.bp.blogspot.com/-0pp9oQtCdX0/UV8BUtg\\\\_6\\\\_I/AAAAAAAAAMk/OVWkD\\\\_jegeM/s1600/goal.jpg)'"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
